File: sock_pairs.go
package main

import "fmt"

func sockMerchant(n int, ar []int) (pairs int) {
	colorsSeen := make(map[int]int)
	for _, sockColorID := range ar {
		_, ok := colorsSeen[sockColorID]
		if !ok {
			colorsSeen[sockColorID] = 1
		} else {
			colorsSeen[sockColorID] += 1
		}
	}

	for _, sockColorAmount := range colorsSeen {
		pairs += sockColorAmount / 2
	}
	fmt.Println(pairs)
	return
}

func main() {
	arr1 := []int{10, 20, 20, 10, 10, 30, 50, 10, 20}
	sockMerchant(len(arr1), arr1)

	arr2 := []int{1, 2, 1, 2, 1, 3, 2}
	sockMerchant(len(arr2), arr2)
}

File: sock_pairs.py
#!/usr/bin/python3
# https://www.hackerrank.com/challenges/sock-merchant/problem


def sockMerchant(n, ar):
    pairs = 0
    colors_seen = dict()
    for sock_color_number in ar:
        if colors_seen.get(sock_color_number) is None:
            colors_seen.update({sock_color_number: 1})
        else:
            colors_seen[sock_color_number] += 1
    for k, v in colors_seen.items():
        pairs += int(v / 2)
    print(pairs)
    return pairs


sockMerchant(9, [10, 20, 20, 10, 10, 30, 50, 10, 20])
sockMerchant(7, [1, 2, 1, 2, 1, 3, 2])
